datafusion/common/src/hash_utils.rs

```diff
@@ -207,6 +208,36 @@ fn hash_dictionary<K: ArrowDictionaryKeyType>(
     Ok(())
 }
 
+fn hash_struct_array(
+    array: &StructArray,
+    random_state: &RandomState,
+    hashes_buffer: &mut [u64],
+) -> Result<()> {
+    let values = array.columns();
+    let nulls = array.nulls();
+    let num_columns = values.len();
+
+    // Skip null columns
+    let valid_indices: Vec<usize> = if let Some(nulls) = nulls {
+        nulls.valid_indices().collect()
+    } else {
+        (0..num_columns).collect()
+    };
+
+    for i in valid_indices {
+        let mut values_hashes = vec![0u64; array.len()];
+        let column_values = values[i].clone();
+        // create hashes for each column
+        create_hashes(&[column_values], random_state, &mut values_hashes)?;
+        let hash = &mut hashes_buffer[i];
```

Comment:
This seems not correct, as `i` here to a column in the struct and `hashes_buffer` to a hash of one value?

Reply:
@Dandandan. Not fully got your point. I do consider each column as one value (iterate each row in one column with get one final hash).

`i` is the column index in the struct array and also is the hashes_buffer index.
For example, column1, null, column3, column4. we iterate 0, 2, and 3. And hash each of them to hashes_buffer[0], hashes_buffer[2], and hashes_buffer[3] respectively. A column-wise hashes.

Reply:
I think I mistake how the null buffer works in StructArray. It seems to be row-wise, but I consider it as column-wise...

Reply:
It looks pretty better now :)